Introduce ModelPresentationManagerProxy that manages the remote model views on the page

achan00 opened the pull request:
#### fce7a7280e2c1e0812936efc57be17d806d49591
<pre>
Introduce ModelPresentationManagerProxy that manages the remote model views on the page
<a href="https://bugs.webkit.org/show_bug.cgi?id=284649">https://bugs.webkit.org/show_bug.cgi?id=284649</a>
<a href="https://rdar.apple.com/141450845">rdar://141450845</a>

Reviewed by Simon Fraser.

Add a new ModelPresentationManagerProxy to manage the presentations of remotely
hosted model contents on the page. ModelPresentationManagerProxy has a map of
model layer identifiers to ModelPresentation objects, which contains info
about the model being hosted (ModelContext), the remote view hosting the model layer,
and the new WKPageHostedModelView which is a WKCompositingView subclass used
in the layer tree node for the model contents. ModelContext is passed
down with the LayerCreationProperties. When setting up the view for the model
contents layer tree node, instead of creating a new WKUIRemoteView with the
layer hosting context ID, it’ll ask ModelPresentationManagerProxy for the view.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::graphicsLayer const):
(WebCore::HTMLModelElement::layerID const):
Helper method to get the PlatformLayerID of the primary layer of the model
element.
(WebCore::HTMLModelElement::modelContentsLayerID const):
Renamed from platformLayerID() so it&apos;s clear which layer it&apos;s for.
(WebCore::HTMLModelElement::modelContext const):
(WebCore::HTMLModelElement::platformLayerID): Deleted.
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Sources.txt
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setContentsToModelContext):
Make this a method dedicated for setting model contents which takes in
a ModelContext.
(WebCore::GraphicsLayer::setContentsToRemotePlatformContext): Deleted.
* Source/WebCore/platform/graphics/ModelContext.cpp
* Source/WebCore/platform/graphics/ModelContext.h
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createPlatformCALayer):
(WebCore::GraphicsLayerCA::setContentsToModelContext):
(WebCore::GraphicsLayerCA::setContentsToRemotePlatformContext): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
Adding ModelContext in the AdditionalData.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::hostingContextID const):
Updated to handle the new ModelContext additional data.
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::modelContext const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.h: Added.
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::layerWillBeRemoved):
If the model element&apos;s layer is being removed, remove the corresponding
ModelPresentation from ModelPresentationManagerProxy.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
Instead of always creating a new WKUIRemoteView, get the model view from ModelPresentationManagerProxy.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
Set up ModelPresentationManagerProxy.
(WebKit::WebPageProxy::resetState):
Invalidate all model presentations on that page.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WKPageHostedModelView.h: Copied from Source/WebCore/Modules/model-element/ModelPlayerClient.h.
* Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm: Copied from Source/WebCore/Modules/model-element/ModelPlayerClient.h.
(-[WKPage:WKPageHostedModelView setRemoteModelView:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::modelPresentationManagerProxy const):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::createPlatformCALayer):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::create):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.mm:
(WebKit::PlatformCALayerRemoteCustom::create):
(WebKit::PlatformCALayerRemoteCustom::PlatformCALayerRemoteCustom):
(WebKit::PlatformCALayerRemoteCustom::populateCreationProperties):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::layerDidEnterContext):

Canonical link: <a href="https://commits.webkit.org/288141@main">https://commits.webkit.org/288141@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b22ba4909e610415b594c8faa46944c6f95d2b5c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81953 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1480 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35910 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86510 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32985 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84059 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9303 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63904 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21630 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85023 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1125 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74574 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44187 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1027 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28751 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31404 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29362 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87941 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9195 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6555 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72272 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9380 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70393 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71494 "Found 1 new API test failure: /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/cut-copy-paste/non-editable (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17826 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15600 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14519 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9146 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14682 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8986 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12511 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10794 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -372,20 +376,34 @@ void HTMLModelElement::didFailLoading(ModelPlayer& modelPlayer, const ResourceEr
-        return std::nullopt;
+        return nullptr;
 
     if (!renderLayerModelObject->isComposited() || !renderLayerModelObject->layer() || !renderLayerModelObject->layer()->backing())
```
This only needs to test `isComposited()`; `isComposited()` tests for the presence of a `backing`.

smfr commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -167,6 +171,11 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
+    RefPtr<GraphicsLayer> graphicsLayer() const;
+    // This returns the PlatformLayerIdentifier of the model element's main layer, which is
+    // different from platformLayerID() that returns the PlatformLayerID for the contents layer.
+    std::optional<PlatformLayerIdentifier> primaryPlatformLayerID() const;
```
Can we resolve this confusion with some different naming? Like `modelContentsLayerID` or something?

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4236,6 +4236,37 @@ void Page::forEachWindowEventLoop(const Function<void(WindowEventLoop&)>& functo
+    if (!frame)
+        return std::nullopt;
+
+    return ElementContext { element.boundingBoxInRootViewCoordinates(), *m_identifier, document->identifier(), element.identifier() };
```
Do all clients need the bounding box? That's quite expensive to compute.

achan00 replied:
This was copied from `WebPage::contextForElement()`. In the new revision of the patch, I'm going to introduce a new `HostedModel` in `WebCore/platform` that has the minimum information we need for now. And this change will be reverted.

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -560,7 +564,9 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
     virtual void setContentsToPlatformLayerHost(LayerHostingContextIdentifier) { }
-    virtual void setContentsToRemotePlatformContext(LayerHostingContextIdentifier, ContentsLayerPurpose) { }
+#if ENABLE(MODEL_PROCESS)
+    virtual void setModelContentsToRemotePlatformContext(const ModelElementContext&, ContentsLayerPurpose) { }
```
The "remote platform" bit doesn't seem necessary and doesn't match the argument's type. And the fact that it's remote doesn't seem relevant at this level.

achan00 replied:
Will address in the next version.

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -79,6 +79,10 @@
 #endif
 
+#if ENABLE(MODEL_PROCESS)
+#include "ModelElementContext.h"
```
This is a layering violation, because ElementContext is in `dom` but here we're in platform code. There is some egregious layering violation here with HTMLMediaElement, but I'd prefer to not propagate that.

achan00 replied:
Will address in the next version.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
```diff
@@ -255,7 +267,7 @@ class RemoteLayerTreeTransaction : public CanMakeCheckedPtr<RemoteLayerTreeTrans
     Markable<WebCore::LayerHostingContextIdentifier> m_remoteContextHostedIdentifier;
 
-    Vector<LayerCreationProperties> m_createdLayers;
+    Vector<UniqueRef<LayerCreationProperties>> m_createdLayers;
```
This might be a perf regression. Can we avoid heap allocating the LayerCreationProperties?

achan00 replied:
Will change this in the next version.

smfr commented on Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.h:
```diff
@@ -0,0 +1,77 @@
+
+    class ModelPresentation {
+        WTF_MAKE_FAST_ALLOCATED;
+    public:
```
Might as well use a `struct` if everything is public.

smfr replied:
You can remove the `public:`

smfr commented on Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm:
```diff
@@ -0,0 +1,106 @@
+{
+}
+
+ModelPresentationManagerProxy::~ModelPresentationManagerProxy()
```
`= default`

smfr commented on Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm:
```diff
@@ -0,0 +1,50 @@
+        return;
+
+    [_remoteModelView removeFromSuperview];
+    [_remoteModelView release];
```
Can we use a RetainPtr<> instead of manual retain/release?

smfr commented on Source/WebCore/platform/graphics/HostedModel.h:
```diff
@@ -0,0 +1,49 @@
+
+namespace WebCore {
+
+class HostedModel final : public RefCounted<HostedModel> {
```
Feels more like ModelContext perhaps? It's just two identifiers, so maybe ref-counting is overkill?

achan00 replied:
We may add more in the future. I mostly followed `WebCore::Model`, and this way we can include in the `LayerCreationProperties` in a `Ref` pointer to keep it from size increase.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
```diff
@@ -97,6 +101,9 @@ class RemoteLayerTreeTransaction : public CanMakeCheckedPtr<RemoteLayerTreeTrans
 #if ENABLE(MODEL_ELEMENT)
             Ref<WebCore::Model>, // PlatformCALayerRemoteModelHosting
+#if ENABLE(MODEL_PROCESS)
+            Ref<WebCore::HostedModel>, // PlatformCALayerRemoteCustom
```
It would be nice not to need both this and the Model.

achan00 replied:
Agreed. I think we need to clean up all of the previous model code (that's before the `MODEL_PROCESS` flag) at some point.

achan00 committed:
Introduce ModelPresentationManagerProxy that manages the remote model views on the page
https://bugs.webkit.org/show_bug.cgi?id=284649
rdar://141450845

Reviewed by Simon Fraser.

Add a new ModelPresentationManagerProxy to manage the presentations of remotely
hosted model contents on the page. ModelPresentationManagerProxy has a map of
model layer identifiers to ModelPresentation objects, which contains info
about the model being hosted (ModelContext), the remote view hosting the model layer,
and the new WKPageHostedModelView which is a WKCompositingView subclass used
in the layer tree node for the model contents. ModelContext is passed
down with the LayerCreationProperties. When setting up the view for the model
contents layer tree node, instead of creating a new WKUIRemoteView with the
layer hosting context ID, it’ll ask ModelPresentationManagerProxy for the view.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::graphicsLayer const):
(WebCore::HTMLModelElement::layerID const):
Helper method to get the PlatformLayerID of the primary layer of the model
element.
(WebCore::HTMLModelElement::modelContentsLayerID const):
Renamed from platformLayerID() so it's clear which layer it's for.
(WebCore::HTMLModelElement::modelContext const):
(WebCore::HTMLModelElement::platformLayerID): Deleted.
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Sources.txt
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setContentsToModelContext):
Make this a method dedicated for setting model contents which takes in
a ModelContext.
(WebCore::GraphicsLayer::setContentsToRemotePlatformContext): Deleted.
* Source/WebCore/platform/graphics/ModelContext.cpp
* Source/WebCore/platform/graphics/ModelContext.h
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createPlatformCALayer):
(WebCore::GraphicsLayerCA::setContentsToModelContext):
(WebCore::GraphicsLayerCA::setContentsToRemotePlatformContext): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
Adding ModelContext in the AdditionalData.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::hostingContextID const):
Updated to handle the new ModelContext additional data.
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::modelContext const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.h: Added.
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::layerWillBeRemoved):
If the model element's layer is being removed, remove the corresponding
ModelPresentation from ModelPresentationManagerProxy.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
Instead of always creating a new WKUIRemoteView, get the model view from ModelPresentationManagerProxy.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
Set up ModelPresentationManagerProxy.
(WebKit::WebPageProxy::resetState):
Invalidate all model presentations on that page.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WKPageHostedModelView.h: Copied from Source/WebCore/Modules/model-element/ModelPlayerClient.h.
* Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm: Copied from Source/WebCore/Modules/model-element/ModelPlayerClient.h.
(-[WKPage:WKPageHostedModelView setRemoteModelView:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::modelPresentationManagerProxy const):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::createPlatformCALayer):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::create):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.mm:
(WebKit::PlatformCALayerRemoteCustom::create):
(WebKit::PlatformCALayerRemoteCustom::PlatformCALayerRemoteCustom):
(WebKit::PlatformCALayerRemoteCustom::populateCreationProperties):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::layerDidEnterContext):

Canonical link: https://commits.webkit.org/288141@main
